### ISSUE TITLE
Error to compile native function with callback arg

### ISSUE THREAD

**Issue by mobilemindtec** [scala-native/scala-native#3740]
Hi,

Can I do create a extern functioon with callback arg? I'm trying, but I get this compilation error:

```scala

type Request = Ptr[CStruct4[CString, CString, CString]]
type Response = Ptr[CStruct5[CInt, CString, CString]]

type Callback = CFuncPtr2[Request, Response, Unit]
type HttpHandlerAsync = CFuncPtr2[Request, Callback, Unit]

@name("run_async")
def runAsync(hostname: CString,
                  port: CUnsignedShort,
                  maxThread: CUnsignedShort,
                  handler: HttpHandlerAsync): CInt = extern

def handlerAsync(req: Request, cb: Callback): Unit = ???

runAsync(c"0.0.0.0", 8081, 1, CFuncPtr2.fromScalaFunction(handlerAsync)) // fail to compile!!

```

The type `HttpHandlerAsync` receive  e `Request` and  a `Callback` which is a function passed by the C code.

```shell
sbt:http-native> run
[info] compiling 1 Scala source to /mnt/data/sources/scala/native/http-native/target/scala-3.3.1/classes ...

  unhandled exception while running scalanative-genNIR on /mnt/data/sources/scala/native/http-native/src/main/scala/Main.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose

     while compiling: <no file>
        during phase: <no phase>
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.10
    compiler version: version 3.3.1
            settings: -Xplugin List(/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/nscplugin_3.3.1/0.4.17/nscplugin_3.3.1-0.4.17.jar) -bootclasspath /home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.3.1/scala3-library_3-3.3.1.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.10/scala-library-2.13.10.jar -classpath /mnt/data/sources/scala/native/http-native/target/scala-3.3.1/classes:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/scala3lib_native0.4_3/0.4.17/scala3lib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/auxlib_native0.4_3/0.4.17/auxlib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/javalib_native0.4_3/0.4.17/javalib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/scalalib_native0.4_2.13/0.4.17/scalalib_native0.4_2.13-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/nativelib_native0.4_3/0.4.17/nativelib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/posixlib_native0.4_3/0.4.17/posixlib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/windowslib_native0.4_3/0.4.17/windowslib_native0.4_3-0.4.17.jar:/home/ricardo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-native/clib_native0.4_3/0.4.17/clib_native0.4_3-0.4.17.jar -d /mnt/data/sources/scala/native/http-native/target/scala-3.3.1/classes

                tree: EmptyTree
       tree position: :<unknown>
           tree type: <notype>
              symbol: val <none>
           call site: package <root> in module class <root>

  == Source file context for tree position ==

[error] ## Exception when compiling 1 sources to /mnt/data/sources/scala/native/http-native/target/scala-3.3.1/classes
[error] scala.scalanative.util.UnsupportedException: can't recover compile-time tag from Apply(Ident(materializeCFuncPtr2),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct4Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct2Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))))))))))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct5Tag),List(Apply(Ident(materializeIntTag),List()), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializeIntTag),List()), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct2Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))))))))))), Apply(Ident(materializeUnitTag),List())))
[error] scala.scalanative.util.UnsupportedException$.apply(UnsupportedException.scala:4)
[error] scala.scalanative.util.package$.unsupported(package.scala:18)
[error] scala.scalanative.nscplugin.NirGenUtil.unwrapTag$$anonfun$1(NirGenUtil.scala:156)
[error] scala.Option.getOrElse(Option.scala:201)
[error] scala.scalanative.nscplugin.NirGenUtil.unwrapTag(NirGenUtil.scala:157)
[error] scala.scalanative.nscplugin.NirGenUtil.unwrapTag$(NirGenUtil.scala:21)
[error] scala.scalanative.nscplugin.NirCodeGen.unwrapTag(NirCodeGen.scala:18)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.$anonfun$71(NirGenExpr.scala:2301)
[error] scala.collection.immutable.List.map(List.scala:250)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genCFuncFromScalaFunction(NirGenExpr.scala:2301)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genApplyPrimitive(NirGenExpr.scala:1048)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genApply(NirGenExpr.scala:111)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genExpr(NirGenExpr.scala:53)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genArg$1(NirGenExpr.scala:1501)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genExternMethodArgs$$anonfun$2(NirGenExpr.scala:1554)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:933)
[error] scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:903)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genExternMethodArgs(NirGenExpr.scala:1555)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genMethodArgs(NirGenExpr.scala:1479)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genApplyMethod(NirGenExpr.scala:1164)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genApplyMethod(NirGenExpr.scala:1146)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genApply(NirGenExpr.scala:117)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genExpr(NirGenExpr.scala:53)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genBlock$$anonfun$1(NirGenExpr.scala:179)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genBlock(NirGenExpr.scala:179)
[error] scala.scalanative.nscplugin.NirGenExpr$ExprBuffer.genExpr(NirGenExpr.scala:55)
[error] scala.scalanative.nscplugin.NirGenStat.genBody$1$$anonfun$2$$anonfun$1(NirGenStat.scala:344)
[error] scala.scalanative.nscplugin.NirGenStat.withOptSynchronized$1(NirGenStat.scala:322)
[error] scala.scalanative.nscplugin.NirGenStat.genBody$1$$anonfun$2(NirGenStat.scala:344)
[error] scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genBody$1(NirGenStat.scala:349)
[error] scala.scalanative.nscplugin.NirGenStat.genMethodBody$$anonfun$1(NirGenStat.scala:355)
[error] scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genMethodBody(NirGenStat.scala:357)
[error] scala.scalanative.nscplugin.NirGenStat.genMethod$$anonfun$1$$anonfun$1(NirGenStat.scala:229)
[error] scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genMethod$$anonfun$1(NirGenStat.scala:236)
[error] scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genMethod(NirGenStat.scala:238)
[error] scala.scalanative.nscplugin.NirGenStat.$anonfun$2(NirGenStat.scala:177)
[error] scala.collection.immutable.List.flatMap(List.scala:293)
[error] scala.scalanative.nscplugin.NirGenStat.genMethods(NirGenStat.scala:182)
[error] scala.scalanative.nscplugin.NirGenStat.genNormalClass(NirGenStat.scala:65)
[error] scala.scalanative.nscplugin.NirGenStat.genClass$$anonfun$1(NirGenStat.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genClass$$anonfun$adapted$1(NirGenStat.scala:47)
[error] scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:46)
[error] scala.scalanative.nscplugin.NirGenStat.genClass(NirGenStat.scala:47)
[error] scala.scalanative.nscplugin.NirGenStat.genClass$(NirGenStat.scala:26)
[error] scala.scalanative.nscplugin.NirCodeGen.genClass(NirCodeGen.scala:18)
[error] scala.scalanative.nscplugin.NirCodeGen.genCompilationUnit$$anonfun$1(NirCodeGen.scala:82)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] scala.scalanative.nscplugin.NirCodeGen.genCompilationUnit(NirCodeGen.scala:82)
[error] scala.scalanative.nscplugin.NirCodeGen.run(NirCodeGen.scala:61)
[error] scala.scalanative.nscplugin.GenNIR.run(GenNIR.scala:18)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:327)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:331)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:246)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:262)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:270)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:279)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:279)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:194)
[error] dotty.tools.dotc.Run.compile(Run.scala:179)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2371)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2321)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:31)
[error] sbt.internal.io.Retry$.apply(Retry.scala:47)
[error] sbt.internal.io.Retry$.apply(Retry.scala:29)
[error] sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2319)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] sbt.Execute.work(Execute.scala:292)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error] java.base/java.lang.Thread.run(Thread.java:1583)
[error]            
[error] stack trace is suppressed; run last Compile / compileIncremental for the full output
[error] (Compile / compileIncremental) scala.scalanative.util.UnsupportedException: can't recover compile-time tag from Apply(Ident(materializeCFuncPtr2),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct4Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct2Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))))))))))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct5Tag),List(Apply(Ident(materializeIntTag),List()), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializeIntTag),List()), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeCStruct2Tag),List(Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))), Apply(Ident(materializePtrTag),List(Apply(Ident(materializeByteTag),List()))))))))))), Apply(Ident(materializeUnitTag),List())))
[error] Total time: 2 s, completed Feb 9, 2024, 10:52:37 AM

```

I managed to solve it using a raw pointer, but I would like to know if there is another way and if this compilation error is expected.

Thanks!

**Comment by WojciechMazur** [comment]
Yes, it should be possible, we have a bug in here. Thank you for a report

**Comment by WojciechMazur** [comment]
This issue was already solved in the 0.5.x branch, the fix for 0.4.x would be published this week
